src/services/factories/getTrades.ts

```diff
@@ -56,14 +62,28 @@ export function getTradesFactory(factoryParams: {
       tokenIdsSet.add(event.sellTokenId)
     })
 
-    // TODO: maybe combine both promise lists into 1?
-    const blockTimes = new Map<number, number>(await Promise.all(Array.from(blocksSet).map(getBlockTimePair)))
-    // TODO: we probably already have this info in global state and don't need to fetch from the contract again?
+    // Filter orders that we might already have
     const orders = new Map<string, Order>(
-      await Promise.all(Array.from(orderIdsSet).map(orderId => getOrderPair(userAddress, networkId, orderId))),
+      existingOrders
+        .filter(order => orderIdsSet.has(order.id))
+        .map(order => {
+          orderIdsSet.delete(order.id)
+          return [order.id, order]
+        }),
     )
-    // TODO: list might not be up to date. Handle case where possibly tokens are not found
-    // In that case, what to do? retry? get token from contract directly? use the `add to local list` functionality?
+    // Fetch from contract the ones we don't have locally
+    const orderPairs = await Promise.all(
+      Array.from(orderIdsSet).map(orderId => getOrderPair(userAddress, networkId, orderId)),
+    )
+    orderPairs.forEach(pair => orders.set(...pair))
+
+    // Fetch block info
+    const blockTimes = new Map<number, number>(await Promise.all(Array.from(blocksSet).map(getBlockTimePair)))
+
+    // Add whatever token might be missing. Will not try to add tokens already in the list.
+    await addUnlistedTokensToUserTokenListById(Array.from(tokenIdsSet))
+
+    // Get the tokens we need
     const tokens = new Map(
       getTokens(networkId)
         .filter(token => tokenIdsSet.has(token.id))
```

Comment:
Why even filter if you just cherry-pick from map later (`tokens.get(event.buyTokenId)`).
Just feed everything into map. Can even do `.forEach` if you want for <sub><sup>micro</sup></sub>optimization

Reply:
Good point :+1: